app/src/main/kotlin/compiler/intermediate_form/Liveness.kt

```diff
@@ -1,10 +1,65 @@
 package compiler.intermediate_form
 
+import compiler.common.reference_collections.ReferenceSet
+import compiler.common.reference_collections.combineReferenceSets
+import compiler.common.reference_collections.referenceHashMapOf
+import compiler.common.reference_collections.referenceHashSetOf
+
 object Liveness {
     data class LivenessGraphs(
         val interferenceGraph: Map<Register, Set<Register>>,
         val copyGraph: Map<Register, Set<Register>>
     )
 
-    fun computeLiveness(linearProgram: List<Asmable>): LivenessGraphs = TODO()
+    object LivenessDataFlowAnalyser : DataFlowAnalyser<ReferenceSet<Register>>() {
+        override val backward = true
+
+        override val entryPointValue: ReferenceSet<Register> = referenceHashSetOf()
+
+        override val latticeMaxElement: ReferenceSet<Register> = referenceHashSetOf()
+
+        override fun latticeMeetFunction(elements: Collection<ReferenceSet<Register>>): ReferenceSet<Register> {
+            return combineReferenceSets(elements.toList())
+        }
+
+        override fun transferFunction(instruction: Instruction, inValue: ReferenceSet<Register>): ReferenceSet<Register> {
+            val gen = instruction.regsUsed
+            val kill = instruction.regsDefined
+            return combineReferenceSets(referenceHashSetOf(gen.toList()), referenceHashSetOf(inValue.filter { it !in kill }))
+        }
+    }
+
+    fun computeLiveness(linearProgram: List<Asmable>): LivenessGraphs {
+        val instructionList = linearProgram.filterIsInstance<Instruction>()
+
+        // find all registers, prepare empty graphs
+        val allRegisters = referenceHashSetOf<Register>()
+        instructionList.forEach {
+            allRegisters.addAll(it.regsDefined)
+            allRegisters.addAll(it.regsUsed)
+        }
+
+        val interferenceGraph = referenceHashMapOf(allRegisters.map { it to referenceHashSetOf<Register>() })
+        val copyGraph = referenceHashMapOf(allRegisters.map { it to referenceHashSetOf<Register>() })
+
+        // run analysis and fill the graphs
+        val dataFlowResult = LivenessDataFlowAnalyser.analyse(linearProgram)
+        val outLiveRegisters = dataFlowResult.outValues.mapValues { (instr, regs) -> combineReferenceSets(regs, referenceHashSetOf(instr.regsDefined.toList())) }
+
+        instructionList.forEach {
+            for (definedReg in it.regsDefined)
+                for (regLiveOnOutput in outLiveRegisters[it]!!)
+                    if (definedReg !== regLiveOnOutput) {
+                        if (it is Instruction.InPlaceInstruction.MoveRR && it.reg_dest == definedReg && it.reg_src == regLiveOnOutput) {
```

Comment:
I'm not 100% sure, but if the instruction is `x := y`, and `y` is never used again, then y is not in `outLiveRegisters[it]!!`, and thus there is no `x--y` edge in the copy graph, but we would like it to be there.

Reply:
You're right, corrected and added a test for that.